### ISSUE TITLE
New BSIP: Market liquidity pool 

### ISSUE THREAD

**Issue by froooze** [bitshares/bsips#213]
# Problems
- User has no easy way to provide liquidity to market pairs
- Most order books have low or medium liquidity 
- Most order books have a high spread
- Order books are dependent on certain price levels

# Solution
- The `market liquidity pool` is based on the [x * y = k modell](https://github.com/runtimeverification/verified-smart-contracts/blob/uniswap/uniswap/x-y-k.pdf) to improve market liquidity.
- Every asset can have a `market liquidity pool` with a BTS/asset market pair
- Adding/removing a asset to/from the `liquidity pool`, only possible when also adding/removing the equivalent amount of BTS to/from the `liquidity pool`

## Fees:
- is paid in BTS and added to the pool to change k
- is only paid for BTS -> Asset trading
- higher than standard market fee

### Examples
- 0.33% BTS->asset market fee
- 20% discount for LTM
- 10% to the reserve pool 
- 70% to the BTS/asset holder   

## Market dynamics
- Every asset is backed with same amount of BTS
- When BTS price decreases, demand for BTS does also increase
- With no spread, every asset can be efficiently traded against each other 
- Liquidity pool does support the order book with extra liquidity and trading pairs
- When order book is limited or exhausted on a price level the `liquidity pool` gets a price advantage

# Sources:
- [Uniswap Whitepaper](https://hackmd.io/C-DvwDSfSxuh-Gd4WKE_ig)
- [x * y = k modell](https://github.com/runtimeverification/verified-smart-contracts/blob/uniswap/uniswap/x-y-k.pdf)
- [Uniswap — A Unique Exchange](https://medium.com/scalar-capital/uniswap-a-unique-exchange-f4ef44f807bf)

**Comment by abitmore** [comment]
@ryanRfox maybe assign 78 to this? Since 77 was assigned to #161

**Comment by ioBanker** [comment]
This will add a huge amount of liquidity and act the same exact way as the centralized exchanges non-stop liquidity bots, it will help the BTS value and run the market dynamics.

**Comment by tbone-bts** [comment]
This looks very promising. I think the combination of this with [BSIP 70](https://github.com/bitshares/bsips/issues/170) could simultaneously help attract both a) more traders and b) more liquidity providers, thus potentially kicking off a sustainable, virtuous cycle of liquidity and adoption. 

Time is always of the essence (of course), but this actually seems more foundational than BSIP 70, so I am eager to see what some of the devs and others familiar with the platform's market mechanics think about it.

**Comment by CryptickCryptick1** [comment]
Copying Uniswap is an interesting concept, but one I did not get until after I read several articles and then saw an Youtube interview with the founder. https://epicenter.tv/episode/292/

Uniswap is not an easy or fast concept to get. But once you get it is a very fascinating concept.

I have long been fascinated with Bancor. Bancor has some problems, but the market maker incentives spread between the liquidity providers forming a type of dividend for those providing liquidity which seems to work well in markets for small traders. It appears that crypto has brought out a number of interesting market making/order matching ideas that never really took off in other markets such as wall street or futures. 

I think this approach needs further study and a better simpler presentation. This would help more widespread review. It might be better to point now comers to a less technical document somewhere.  It would be a fundamental change in how the market works. The exchange using the concept is less that one year old and there are some people who now appearing to discovering and utilizing various strategies to optimize earnings and potentially exploiting a few loop holes.

The atmosphere in BitShares may not be right for serious consideration of the strategy now. (BitAsset settlement and the upvoting of refund 400K has stalled progress.) 

I look forward to seeing what other traders think of this exciting concept.

**Comment by froooze** [comment]
@sschiessl-bcp

**Comment by sschiessl-bcp** [comment]
> @sschiessl-bcp

Have you changed anything in here?

**Comment by froooze** [comment]
> Have you changed anything in here?

No, he wanted just to get pinged here!

**Comment by froooze** [comment]
@ryanRfox 
Can you comment for assigning this BSIP ?

**Comment by matle85** [comment]
Anything that increases liquidity gets the thumbs up from me - not familiar with uniswap so not 100% clear on how it works at first reading. 

Basically funds from multiple assets can be put in a shared liquidity pool and then the blockchain acts in the background as a sort of trustless bot / liquidity provider in the background providing swaps between the various assets that have contributed?

I'm a visual guy, a sketch would help 🙈

**Comment by froooze** [comment]
> Basically funds from multiple assets can be put in a shared liquidity pool and then the blockchain acts in the background as a sort of trustless bot / liquidity provider in the background providing swaps between the various assets that have contributed?

Yes, any `asset` can have a `MLP` and get a market pair with BTS, except BTS itself. The user can decide, which asset is supported with extra liquidity from `MLP`.

The market price dictates the exchange rate and the ratio of amount_BTS and amount_asset.

**Comment by sschiessl-bcp** [comment]
I'm still missing explanatory text for people unfamiliar with the referenced concepts (like me). In particular, Abstract and Summary should provide more detail and be in prosa.

**Comment by pmconrad** [comment]
I agree that the concept is difficult to understand without explanations in prose.

You should add a section that describes the effect on funds invested into the pool. As long as the assets oscillate around a more or less fixed base price everything is fine - investors participate in the fee income. As the trade price moves further away from the original investment price, the participation in the price appreciation (or loss) gets smaller and smaller. This may be seen as a pro or a con, depending on investment goals. :-) (For a pool consisting of assets A and B, and price p=B/A, the total value in terms of B is equal to SQRT(k*p), i. e. the total value doubles if the price quadruples.)

This also means that when BTS is in a downtrend against USD for example, then investors would probably remove their funds and then sell for USD to cut their losses. Similarly, if BTS is in an uptrend, investors would also withdraw their funds but sell USD for BTS to participate in BTS appreciation. It is therefore doubtful if the pool would help liquidity when it's needed the most.

A big bonus of this protocol over Bancor is that it works without costly `log` and `exp` computations.

**Comment by froooze** [comment]
> (For a pool consisting of assets A and B, and price p=B/A, the total value in terms of B is equal to SQRT(k*p), i. e. the total value doubles if the price quadruples.)

Yes, `pool_value` ~ sqrt(price_BTS), had to correct some equations. Should we allow asset pairs without BTS as base pair? This would bring some flexibility. 

> This also means that when BTS is in a downtrend against USD for example, then investors would probably remove their funds and then sell for USD to cut their losses. Similarly, if BTS is in an uptrend, investors would also withdraw their funds but sell USD for BTS to participate in BTS appreciation. It is therefore doubtful if the pool would help liquidity when it's needed the most.

Yes, this is always possible, but this concept is more for user, who don't want to trade or setup a DEXbot worker, but earn passive income. Most funds on the blockchain are not on the orderbook, but just lay around and do nothing. 

> A big bonus of this protocol over Bancor is that it works without costly `log` and `exp` computations.

This and no need to set/use the connector weight parameter.

**Comment by matle85** [comment]
Thanks  - sounds like it would be a good addition and a nice method for people to assist in providing liquidity without running dexbot.

Have the team estimated the time for implementation? Maybe we can get a few people together committing to try it with some funds once its live?

**Comment by matle85** [comment]
Re: the discussion on base. I think we need to have a big focus on the tokenomics of bitshares moving forward so I think BTS should be at the centre of every new initiative.

On that basis I'd view the ideal solution being BTS as the base / background behind all asset pairs.

**Comment by froooze** [comment]
MLP version updated!

**Comment by abitmore** [comment]
If it's an "OIP" then I would like to close it. If it's a BSIP then please update wording in the interest of BitShares.
